cmd/consul-dataplane/merge.go

```diff
@@ -0,0 +1,269 @@
+package main
+
+import (
+	"strings"
+
+	"github.com/hashicorp/consul-dataplane/pkg/consuldp"
+)
+
+// mergeConfigs takes in a couple of configs(c1&c2) and applies
+// c2 on to c1 based on the following conditions
+//
+// 1. If an attribute of c2 is a string/integer/bool, then it is applied to the
+// same field in c1 if it holds non default value for the field
+// 2. If an attribute of c2 is a slice/map, then it overrides the same field in c1
+func mergeConfigs(c1, c2 *consuldp.Config) {
+	mergeConsulConfigs(c1.Consul, c2.Consul)
+	mergeServiceConfigs(c1.Service, c2.Service)
+	mergeTelemetryConfigs(c1.Telemetry, c2.Telemetry)
+	mergeLoggingConfigs(c1.Logging, c2.Logging)
+	mergeEnvoyConfigs(c1.Envoy, c2.Envoy)
+	mergeXDSServerConfigs(c1.XDSServer, c2.XDSServer)
+	mergeDNSServerConfigs(c1.DNSServer, c2.DNSServer)
+}
+
+func mergeConsulConfigs(c1, c2 *consuldp.ConsulConfig) {
+	if c2 == nil {
+		return
+	}
+
+	if c2.Addresses != "" {
+		c1.Addresses = c2.Addresses
+	}
+
+	if c2.GRPCPort != int(0) && c2.GRPCPort != DefaultGRPCPort {
+		c1.GRPCPort = c2.GRPCPort
+	}
+
+	if c2.ServerWatchDisabled {
```

Comment:
lets say in config file `ServerWatchDisabled` was true and in command line flag it was false. This code will not override the flag.

Reply:
I am aware of this case and looking for ways to solve it. It stems from the fact that the flag inputs assume a default value when it is not given via the inputs by the user. Let's say we have the following config

```json
"telemetry": {
   "useCentralConfig": false
}
```

and the user gives the same input via the flag

```
consul-dataplane --use-central-config true --addr abc.dc1
```

In this case, we would expect the `central-config` to be true but it becomes false per the current design. Now let's say the user provided the following input

```
consul-dataplane --addr abc.dc1
```

In this scenario, `central-config` assumes a default value of `true` because the user hasn't provided it via the inputs but the end result has to be `false` because we should respect the value given via the config file. There isn't a clean way to differentiate if a value for a flag is given by the user (user gives the default value) or is the default value added by the system.

Reply:
Will it be better to we can make command line flag mandatory and not assume default values?
That way we always override the value in config file to command line value.

Reply:
also if somehow we differentiate that is it default value or given by user. Then it will solve our problem.
otherwise another solution is we can keep a number instead of boolean 
0 = default
1 = true
2 = false
just sharing possibility, but needs dicussion.

Reply:
> Will it be better to we can make command line flag mandatory and not assume default values?

This will be a breaking change with the current way clients are using dataplane. We should probably look into this.

> otherwise another solution is we can keep a number instead of boolean

There are also string based input fields. So it doesn't seem to be trivial.

One simple way to approach this change is to allow either passing values by flag or by config file. Since dataplane is an internal component we could start with that model for simplicity. I want to hear from reviewers as well around this.

Reply:
This should be fixed. I modified the whole approach to make stuff work.